---
title: "LACHESIS - Analyzing tumor evolution from whole genome sequencing data"
date: "`r format(Sys.Date(), '%m/%d/%Y')`"
abstract: >
  Understanding tumor evolution, particularly the temporal order of driver mutations, is increasingly important for improving diagnosis and treatment strategies. A recent study by Körber et al. (Nature Genetics, 2023) identified a bimodal distribution in the timing of the most recent common ancestor (MRCA) in neuroblastomas. This early or late MRCA distinction strongly correlated with clinical outcomes, suggesting its potential as a prognostic biomarker. To facilitate the application of these findings, we developed a package for estimating MRCA timing using the molecular clock of single-strand somatic nucleotide variant (SSNV) accumulation. The package is based on the statistical model introduced by Körber et al. and provides tools for analyzing tumor evolution. This vignette details the functionality of the package and demonstrates its application through typical workflows, enabling users to integrate MRCA timing into their analyses.
  LACHESIS package version: `r packageVersion("LACHESIS")`

output:
  rmarkdown::html_document:
    highlight: pygments
    toc: true
    toc_float:
      toc_collapsed: true
    toc_depth: 3
    number_sections: false
    fig_width: 5
  bibliography: library.bib
  vignette: >
    %\VignetteIndexEntry{LACHESIS - Analyzing tumor evolution from whole genome sequencing data}
    %\VignetteEngine{knitr::rmarkdown}
    %\VignetteEncoding{UTF-8}
    %\usepackage[utf8]{inputenc}
---

```{r setup, echo=FALSE, results="hide"}
knitr::opts_chunk$set(tidy = FALSE,
                      cache = FALSE,
                      dev = "png",
                      message = FALSE, error = FALSE, warning = TRUE)
```	

# Installation

To install this package, start R (version "4.4") and enter:

```{r install, eval=FALSE}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("LACHESIS")
```
For older versions of R, please refer to the appropriate [Bioconductor release.](https://bioconductor.org/about/release-announcements/)

# Citation

If you use LACHESIS in published research, please cite:

  *Körber V, Stainczyk SA, Kurilov R, Henrich KO, Hero B, Brors B, Westermann F, Höfer T. (2023)
  Neuroblastoma arises in early fetal development and its evolutionary duration predicts outcome.
  *Nat Genet.*, **55**:619-630.
  [10.1038/s41588-023-01332-y](https://www.nature.com/articles/s41588-023-01332-y)*

Other Bioconductor packages with similar aims are
[name](link).

# Input Data

The input for LACHESIS can be specified by vectors or in a tab-delimited **sample-specification file**.
This file must contain the sample name, path to the SNV file, path to CNV file, and optionally further specifications. 
While executing LACHESIS, a detailed **log file** will be generated, capturing the date, time, and provided input parameters.

#### CNV File

A CNV file describes **Copy Number Variations** (CNVs), which are genomic alterations that result in abnormal copies of one or more genes, caused by duplications, deletions, translocations, and inversions. The CNV file requires columns for the **chromosome number**, **start** and **end position** of the segment, and either the **total copy number** or the **number of A- and B-alleles**. 

#### SNV File

A SNV file describes **Single Nucleotide Variants** (SNVs), which are variations of one nucleotide at a specific location in the genome. The SNV file should be in VCF format (Variant Call Format), a standardized text file format with a **header section** (lines starting with # and ##) and a **data section** where each row represents a variant. 

# Package Overview

LACHESIS utilizes whole genome sequencing data to determine the mutational timing of the MRCA and the ECA. The output includes a folder for each patient ID, which contains the following:

#### Datatables (.txt)

* **MRCA_densities_ID:** chromosomal location, total and subclonal copy numbers (A and B), segment length, mutation counts, mutation densities with confidence intervals, probabilities of segments tracing back to the MRCA or ECA, and inferred evolutionary timing for A and B subclones.
* **SNV_timing_per_segment_ID:** purity, ploidy, and the estimated mean, lower, and upper bounds for the timing of the MRCA and ECA.

#### Graphs (.pdf)

* **SNV_densities:** histograms of mean mutation densities and timeline of early tumor evolution
* **VAF_histogram_strat:** plot of measured copy numbers along the genome and variant allele frequency (VAF) histograms of SNVs stratified by copy number and minor/major allele count
* **VAF_histogram:** plot of frequency distribution of variant allele frequencies

Additionally, if the function was applied to a cohort, the following files will be included:

* **SNV_densities_cohort:** histograms of mean mutation densities and cumulative distribution of mean mutation densities with 95% confidence intervals
* **lachesis_clin.par:** correlation between SNV densities computed by LACHESIS and clinical parameters (age, survival,...)
* **survival_plots:** will be added 

## Structure

There are two ways to perform analyses using LACHESIS: either by utilizing the sub-functions individually or by calling the wrapper function. All of the functions presented in the main workflow are also invoked within the wrapper function. In addition, the generated data is merged, allowing cohort analysis to be performed.

```{r, echo=FALSE, fig.show='hold', out.width="45%"}
knitr::include_graphics(system.file("extdata", "package_structure_main.png", package = "LACHESIS"))

knitr::include_graphics(system.file("extdata", "package_structure_wrapper.png", package = "LACHESIS"))
```

# LACHESIS Main Workflow

## Importing Data

### Importing Copy Number Information (readCNV)

The readCNV function converts a user-specified BED file containing copy number information into a standardized format. It performs several quality checks on the input file and returns a clean, structured dataframe. 
Users can specify column identifiers for chromosomal positions and allele-specific copy number (A and B alleles) using either column names or indices. If no identifiers are provided, the function will attempt to identify the relevant columns using standard nomenclature (e.g., "chrom", "start", "end", etc.). 

#### Handling Missing Allele Information
If total copy number information is available but the number of A and B alleles is missing, the function estimates allele counts. The total copy number is divided by two, with one allele rounded up and the other rounded down.

#### Input Parameters 

```
readCNV <- function(cn.info = NULL, chr.col = NULL, start.col = NULL, end.col = NULL, A.col = NULL, B.col = NULL, tcn.col = NULL, merge.tolerance = 10^5, ignore.XY = TRUE, max.cn = 4, tumor.id = NULL)
```

#### Example: ACESeq

```{r, message=FALSE, warning=FALSE}
aceseq_cn = system.file("extdata", "ACESeq/NBE11_comb_pro_extra2.59_0.83.txt", package = "LACHESIS")
cn_data = LACHESIS::readCNV(aceseq_cn)
head(cn_data, n = 1)
```

#### Example: ASCAT

```{r, message=FALSE, warning=FALSE}
ascat_cn = system.file("extdata", "ASCAT/S98.segments.txt", package = "LACHESIS")
cn_data = LACHESIS::readCNV(ascat_cn)
head(cn_data, n = 1)
```

### Importing Variant Information (readVCF)  

The readVCF function is used to import a VCF file and extract key genomic information such as read counts and VAFs. Supported tools to generate the VCF file are [Mutect2](https://gatk.broadinstitute.org/hc/en-us/articles/360037593851-Mutect2), [Strelka2](https://github.com/Illumina/strelka?tab=readme-ov-file), [Sentieon](https://support.sentieon.com/manual/) and DKFZ.

#### Input Parameters

```
readVCF = function(vcf = NULL, ignore.XY = TRUE, vcf.source = "strelka", min.vaf = 0.01, min.depth = 30, t.sample = NULL, info.af = "AF", info.dp = "DP")
```

#### Example: mutect

```{r, message=FALSE, warning=FALSE}
mutect_vcf = system.file("extdata", "mutect.somatic.vcf.gz", package = "LACHESIS")
m_data = LACHESIS::readVCF(vcf = mutect_vcf, vcf.source = "mutect")
head(m_data, n = 1)
```

#### Example: strelka

```{r, message=FALSE, warning=FALSE}
strelka_vcf = system.file("extdata", "strelka2.somatic.snvs.vcf.gz", package = "LACHESIS")
s_data = LACHESIS::readVCF(vcf = strelka_vcf, vcf.source = "strelka")
head(s_data, n = 1)
```

#### Example: dkfz

```{r, message=FALSE, warning=FALSE}
dkfz_vcf = system.file("extdata", "NBE15", "snvs_NBE15_somatic_snvs_conf_8_to_10.vcf", package = "LACHESIS")
d_data = LACHESIS::readVCF(vcf = dkfz_vcf, vcf.source = "dkfz")
head(d_data, n = 1)
```


### Plotting Variant Allele Frequencies (PlotVAFdistr) 

The plotVAFdistr function generates a histogram of VAFs and a density plot if the showdensity parameter is set to TRUE.

#### Input Parameters 

```
plotVAFdistr(vaf = NULL, vafbreak = 0.05, t_sample = NULL, showcounts = FALSE, showdensity = TRUE, col = "#34495e", srtcounts = 45, output.file = NULL) 
```

#### Example 1

```{r, message=FALSE, warning=FALSE, fig.cap="Figure 1: VAF distribution of SSNVs"}
strelka_vcf = system.file("extdata", "strelka2.somatic.snvs.vcf.gz", package = "LACHESIS")
s_data = LACHESIS::readVCF(vcf = strelka_vcf, vcf.source = "strelka")
LACHESIS::plotVAFdistr(s_data)
```

### Merging Imported Data (NbImport)

The function nbImport merges CNVs and SNVs into a single datatable. Each variant is assigned to its corresponding copy number segment and status. 

#### Input Parameters 

```
nbImport <- function(cnv = NULL, snv = NULL, purity = NULL, ploidy = NULL) 
```

#### Example 1

```{r, message=FALSE, warning=FALSE}
snvs <- system.file("extdata", "NBE15", "snvs_NBE15_somatic_snvs_conf_8_to_10.vcf", package = "LACHESIS")
s_data <- LACHESIS::readVCF(vcf = snvs, vcf.source = "dkfz")
aceseq_cn <- system.file("extdata", "NBE15", "NBE15_comb_pro_extra2.51_1.txt", package = "LACHESIS")
c_data <- LACHESIS::readCNV(aceseq_cn)
nb <- LACHESIS::nbImport(cnv = c_data, snv = s_data, purity = 1, ploidy = 2.51)
head(nb, n = 1)
```

### Plotting Imported Data (plotNB)  

This function visualizes the results generated by nbImport. The top plot displays the measured copy numbers across the genome, each chromosome is represented along the x-axis, with the corresponding copy number values on the y-axis. By default, the human genome build hg19 is used as the reference. However, alternative genome builds such as hg18 or hg38 can be specified if required.
The bottom plots focus on VAF histograms of SNVs. These histograms are stratified based on copy number as well as the counts of the minor and major alleles. 

#### Input Parameters

```
plotNB(nb = NULL, ref_build = "hg19", min.cn = 2, max.cn = 4, samp.name = NULL, output.file = NULL)
```

```{r, message=FALSE, warning=FALSE, fig.cap="Figure 2: *top plot* - copy number profile along the genome, *bottom plots* - VAF distribution of SSNVs stratified by copy number and minor/ major alleles"}
snvs <- system.file("extdata", "NBE15", "snvs_NBE15_somatic_snvs_conf_8_to_10.vcf", package = "LACHESIS")
s_data <- LACHESIS::readVCF(vcf = snvs, vcf.source = "dkfz")
aceseq_cn <- system.file("extdata", "NBE15", "NBE15_comb_pro_extra2.51_1.txt", package = "LACHESIS")
c_data <- LACHESIS::readCNV(aceseq_cn)
nb <- LACHESIS::nbImport(cnv = c_data, snv = s_data, purity = 1, ploidy = 2.51)
LACHESIS::plotNB(nb)
```

## Processing Clonal Mutations 

### Counting Clonal Mutations (ClonalMutationalCounter)

This function counts the number of clonal mutations residing on a single or multiple copies per genomic segment. Segments of equal copy number and B-allele count are merged per chromosome. 

#### Input Parameters 

```
clonalMutationCounter <- function(nbObj = NULL, min.cn = 1, max.cn = 4, chromosomes = c(1:22)) 
```

#### Example 1

```{r, message=FALSE, warning=FALSE}
snvs <- system.file("extdata", "NBE15", "snvs_NBE15_somatic_snvs_conf_8_to_10.vcf", package = "LACHESIS")
s_data <- LACHESIS::readVCF(vcf = snvs, vcf.source = "dkfz")
aceseq_cn <- system.file("extdata", "NBE15", "NBE15_comb_pro_extra2.51_1.txt", package = "LACHESIS")
c_data <- LACHESIS::readCNV(aceseq_cn)
nb <- LACHESIS::nbImport(cnv = c_data, snv = s_data, purity = 1, ploidy = 2.51)
cl_muts <- LACHESIS::clonalMutationCounter(nb)
head(cl_muts, n = 1)
```

### Normalizing Clonal Mutation Counts (normalizeCounts)

This function normalizes the clonal mutation counts obtained with clonalMutationCounter. The normalized counts correspond to the number of mutations accumulated between conception/gastrulation and MRCA/copy number gain. They can hence be interpreted as "molecular time". 

#### Input Parameters 

```
normalizeCounts <- function(countObj)
```

#### Example 1

```{r, message=FALSE, warning=FALSE}
snvs <- system.file("extdata", "NBE15", "snvs_NBE15_somatic_snvs_conf_8_to_10.vcf", package = "LACHESIS")
s_data <- LACHESIS::readVCF(vcf = snvs, vcf.source = "dkfz")
aceseq_cn <- system.file("extdata", "NBE15", "NBE15_comb_pro_extra2.51_1.txt", package = "LACHESIS")
c_data <- LACHESIS::readCNV(aceseq_cn)
nb <- LACHESIS::nbImport(cnv = c_data, snv = s_data, purity = 1, ploidy = 2.51)
cl_muts <- LACHESIS::clonalMutationCounter(nb)
norm_muts <- LACHESIS::normalizeCounts(cl_muts)
head(norm_muts, n = 1)
```

## Inferring Clonal Evolution

### Estimating Mutation Densities at ECA/ MRCA (MRCA)

This function takes the normalized clonal mutation counts obtained with normalizeCounts to estimate mutation densities at MRCA and ECA. Optionally, the average false positive rate of clonal mutations (e.g. due to incomplete tissue sampling) and the standard deviation of the false positive rate can be defined (fp.mean, fp.sd).

#### Input Parameters 

```
MRCA <- function(normObj = NULL, min.seg.size = 10^7, fp.mean = 0, fp.sd = 0, excl.chr
```
#### Example 1

```{r, message=FALSE, warning=FALSE}
snvs <- system.file("extdata", "NBE15", "snvs_NBE15_somatic_snvs_conf_8_to_10.vcf", package = "LACHESIS")
s_data <- LACHESIS::readVCF(vcf = snvs, vcf.source = "dkfz")
aceseq_cn <- system.file("extdata", "NBE15", "NBE15_comb_pro_extra2.51_1.txt", package = "LACHESIS")
c_data <- LACHESIS::readCNV(aceseq_cn)
nb <- LACHESIS::nbImport(cnv = c_data, snv = s_data, purity = 1, ploidy = 2.51)
cl_muts <- LACHESIS::clonalMutationCounter(nb)
norm_muts <- LACHESIS::normalizeCounts(cl_muts)
mrca <- LACHESIS::MRCA(norm_muts) 
head(mrca, n = 1)
```

### Plotting Mutation Densities (plotMutationalDensities)

The figure visualizes the results of the MRCA analysis.
The top plot presents histograms of mean mutation densities, while the bottom plot illustrates a timeline of early tumor evolution. This timeline highlights mutation densities (mean values and 95% confidence intervals) associated with individual chromosomal gains. Additionally, the mutation densities at the ECA and MRCA are depicted. 

#### Input Parameters

```
plotMutationDensities <- function(mrcaObj = NULL, samp.name = NULL, min.seg.size = 10^7, fill.zero = NULL, fill.multi = NULL, l.col = NULL, show.den = NULL, bins = NULL, output.file = NULL)
```
#### Example 1

```{r, message=FALSE, warning=FALSE, fig.cap="Figure 3: *top plots* - Densities of non-amplified and amplified clonal mutations, *bottom plot* - Mutation densities of ECA and MRCA"}
snvs <- system.file("extdata", "NBE15", "snvs_NBE15_somatic_snvs_conf_8_to_10.vcf", package = "LACHESIS")
s_data <- LACHESIS::readVCF(vcf = snvs, vcf.source = "dkfz")
aceseq_cn <- system.file("extdata", "NBE15", "NBE15_comb_pro_extra2.51_1.txt", package = "LACHESIS")
c_data <- LACHESIS::readCNV(aceseq_cn)
nb <- LACHESIS::nbImport(cnv = c_data, snv = s_data, purity = 1, ploidy = 2.51)
cl_muts <- LACHESIS::clonalMutationCounter(nb)
norm_muts <- LACHESIS::normalizeCounts(cl_muts)
mrca <- LACHESIS::MRCA(norm_muts) 
LACHESIS::plotMutationDensities(mrca)
```

# LACHESIS Wrapper Function

This function calls all of the previously mentioned functions and saves the results to the specified output directory.

#### Input Parameters 

```
LACHESIS <- function(input.files = NULL, ids = NULL, vcf.tumor.ids = NULL, cnv.files = NULL, snv.files = NULL, 
                      vcf.source = NULL, purity = NULL, ploidy = NULL, cnv.chr.col = NULL, cnv.start.col = NULL, 
                      cnv.end.col = NULL, cnv.A.col = NULL, cnv.B.col = NULL, cnv.tcn.col = NULL, age = NULL,
                      OS.time = NULL, OS = NULL, EFS.time = NULL, EFS = NULL, output.dir = NULL, 
                      ignore.XY = TRUE, min.cn = 1, max.cn = 4, merge.tolerance = 10^5, min.vaf = 0.01, 
                      min.depth = 30, vcf.info.af = "AF", vcf.info.dp = "DP", min.seg.size = 10^7, fp.mean = 0, 
                      fp.sd = 0, excl.chr = NULL, ref_build = "hg19", ...)
```

#### Example with vectors

```{r, message=FALSE, warning=FALSE}
#an example file with sample annotations and meta data
input.files = system.file("extdata", "Sample_template.txt", package = "LACHESIS")
input.files = data.table::fread(input.files)

#cnv and snv files for example tumors
nbe11 = list.files(system.file("extdata/NBE11/", package = "LACHESIS"), full.names = TRUE)
nbe15 = list.files(system.file("extdata/NBE15/", package = "LACHESIS"), full.names = TRUE)
nbe63 = list.files(system.file("extdata/NBE63/", package = "LACHESIS"), full.names = TRUE)

cnv.file = c(nbe11[1], nbe15[1], nbe63[1])
snv.file = c(nbe11[2], nbe15[2], nbe63[2])

input.files$cnv.file = cnv.file
input.files$snv.file = snv.file

# Make an example input file with paths to cnv and snv file along with other meta data
lachesis_input = tempfile(pattern = "lachesis", tmpdir = tempdir(), fileext = ".tsv")
data.table::fwrite(x = input.files, file = lachesis_input, sep = "\t")

#Example with template file with paths to multiple cnv/snv files as an input
lachesis <- LACHESIS::LACHESIS(input.files = lachesis_input)

head(lachesis, n=1)
```

#### Example with tab-delimited sample-specification file

```{r, message=FALSE, warning=FALSE}
input.files = system.file("extdata", "Sample_template.txt", package = "LACHESIS")
lachesis <- LACHESIS::LACHESIS(input.files = lachesis_input)

head(lachesis, n=1)
```

### Plotting Cohort Analysis (plotLACHESIS)

The plotLACHESIS function visualizes the cumulative SSNV densities of the ECA and MRCA.

#### Input Parameters 

``` 
plotLachesis(lachesis = NULL, suppress.outliers = FALSE, log.densities = FALSE, fill.multi = NULL, l.col = NULL, binwidth = NULL, fill.zero = NULL, output.file = NULL)
```

#### Example 1

```{r, message=FALSE, warning=FALSE, fig.cap="Figure 4: Cumulative SSNV densities of the ECA (dark green) and MRCA (light green)"}
input.files = system.file("extdata", "Sample_template.txt", package = "LACHESIS")
lachesis <- LACHESIS::LACHESIS(input.files = lachesis_input)

LACHESIS::plotLachesis(lachesis)
```

### Plotting Clinical Correlations (plotClinicalCorrelations)

This function takes SNV densities as input and correlates them with clinical data specified by the user, such as age at diagnosis, survival data etc.

#### Input Parameters

```
plotClinicalCorrelations(lachesis = NULL, clin.par = "Age", suppress.outliers = FALSE, log.densities = FALSE, output.file = NULL)
````

#### Example 1

```{r, message=FALSE, warning=FALSE, fig.cap="Figure 5: Correlation between Age and SNV densities at ECA and MRCA"}
input.files = system.file("extdata", "Sample_template.txt", package = "LACHESIS")
lachesis <- LACHESIS::LACHESIS(input.files = lachesis_input)
LACHESIS::plotClinicalCorrelations(lachesis)

```

### Plotting Survival Analysis (plotSurvival)

*will be added*

# How To Get Help 

In case of any questions, feel free to open an issue on [**Github**.](https://github.com/VerenaK90/LACHESIS_shiny/issues) 

Alternatively, questions can be posted to [**Bioconductor support site**](https://support.bioconductor.org), tagging the question with "LACHESIS" will automatically send an alert to the package authors to respond on the support site.  

# Acknowledgments

# Funding


# Session Info

```{r sessionInfo}
sessionInfo()
```

# References
