#' Imports single-nucleotide-variant and copy-number information and combines them into a joint evolve.obj
#' @description
#' Imports a VCF file and extracts read-count and variant allele frequencies, then converts a user-specified bed-file with copy number
#' information into a standardized format. The function performs various quality checks on the file input and combines SNV and CNV information into a
#' `SummarizedExperiment`container.
#'
#' @param vcf Input indexed VCF file. Currently VCF files generated by `mutect2` and `strelka2` are supported.
#' @param cn.info Path to the copy number information. Requires columns for the chromosome number, start and end of the segment, and either the total copy number or the number of A- and B-alleles.
#' @param t.sample Sample name for tumor. Must be same as in VCF. Strelka hardcodes tumor sample name to "TUMOR".
#' @examples
#' mutect_vcf = system.file("extdata", "mutect.somatic.vcf.gz", package = "NBevolution")
#' strelka_vcf = system.file("extdata", "strelka2.somatic.snvs.vcf.gz", package = "NBevolution")
#' aceseq_cn = system.file("extdata", "ACESeq/NBE11_comb_pro_extra2.59_0.83.txt", package = "NBevolution")
#' ascat_cn = system.file("extdata", "ASCAT/S98.segments.txt", package = "NBevolution")
#' nbevolve_mac <- nbImport(mutect_vcf, aceseq_cn, vcf.source = "mutect")
#' @importFrom utils read.delim
#' @export
#' @import data.table vcfR
#' @return a SummarizedExperiment containing a data.table with chrom, pos, ref, alt, t_ref_count, t_alt_count, t_depth, t_vaf and a tandardized data frame with copy number information per segment.
#' @export


nbImport <- function(vcf = NULL, cn.info = NULL, t.sample = NULL, ...){
  # read in the VCF information
  vcf <- readVCF(vcf = vcf, t.sample = t.sample, ...)
  # translate this information into a summarized experiment
  rownames(vcf) <- paste(vcf$chrom, vcf$pos, vcf$alt, sep=".")
  vcf.se <- SummarizedExperiment::SummarizedExperiment(assays = list(vcf))
  # read in the CNV information
  cnv <- readCNV(cn.info = cn.info, ...)
  # generate a summarized experiment
  evolve.obj <- MultiAssayExperiment::MultiAssayExperiment(assays = list(SNVs = vcf, CNV = cnv))
}
